Extract a method that filters frames in the captured stack trace

cjiang1209 opened the pull request:
Summary: The subclass can override the filtering logic to customize which frames to keep or drop.

Test Plan:
```
buck run caffe2/test:test_export -- -r  test_stack_trace
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:others -- -r test_constant_random
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r test_custom_obj_list_out
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:fx  -- -r class_member_back_compat
```


cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

ezyang reviewed (changes requested):
Can you add more comments and oss tests? No longer lear that eliminating private method is bc breaking

ezyang reviewed (changes requested):
you need to reexport

cjiang1209 commented:
> you need to reexport

Yeah the export did not work out. Will try again.

cjiang1209 commented:
> Can you add more comments and oss tests? No longer lear that eliminating private method is bc breaking

Regarding oss tests, there exist some.

https://github.com/pytorch/pytorch/blob/main/test/export/test_export.py#L11808

ezyang commented:
add urself to metamates team bunnylol oss pytorch/pytorch so you don't need ci approvals

ezyang commented on test/export/test_export.py:
```diff
@@ -11976,6 +11977,30 @@ def forward(self, x):
+        graph = TestTracer().trace(Foo())
+
+        trace_x = [node for node in graph.nodes if node.name == "x"][0].stack_trace
+        self.assertTrue(re.search(r"torch/fx/proxy.py.*in create_node\n", trace_x))
```
this probably doesn't work on windows

cjiang1209 committed:
Extract a method that filters frames in the captured stack trace (#158266)

Summary:

The subclass can override the filtering logic to customize which frames to keep or drop.

Test Plan:
```
buck run caffe2/test:test_export -- -r  test_stack_trace -r test_filter_traceback_frames
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:others -- -r test_constant_random
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r test_custom_obj_list_out
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:fx  -- -r class_member_back_compat
```

Reviewed By: ezyang, yushangdi

Differential Revision: D78030234
